# Dataset Heart

Some classification problem on heart.csv dataset

In [1]:
import os
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, PowerTransformer
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB

In [2]:
directory = os.path.join('D:/Bootcamp ML - Mada/classification dataset/', 'heart.csv')
if os.path.isfile(directory):
  print("File ditemukan")
else:
    print("tidak ada")

File ditemukan


In [3]:
dataset = pd.read_csv(directory)
dataset

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [4]:
dataset.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


## Correlation matrix of the features

In [6]:
correlation_matrix = dataset.corr()
corr = correlation_matrix['target'].sort_values(ascending=False)
correlation_dataframe = pd.DataFrame({'column': corr.index,
                 'Correlation with target': corr.values})
correlation_dataframe

column  Correlation with target
0     target                 1.000000
1         cp                 0.433798
2    thalach                 0.421741
3      slope                 0.345877
4    restecg                 0.137230
5        fbs                -0.028046
6       chol                -0.085239
7   trestbps                -0.144931
8        age                -0.225439
9        sex                -0.280937
10      thal                -0.344029
11        ca                -0.391724
12   oldpeak                -0.430696
13     exang                -0.436757

# Feature Engineering

In [7]:
label = dataset['target'] #--> Target Prediksi
feature_used = dataset.drop(['target',],axis=1) #--> Fitur yang digunakan adalah selain fitur 'target'

# Menggunakan 2 metode scalling, Standard dan Power Transform dengan yeo-jhonson
scaler_1 = StandardScaler(with_std=True,with_mean=True)
scaler_2 = PowerTransformer(method='yeo-johnson', standardize=True)

x_1 = scaler_1.fit_transform(feature_used)
x_2 = scaler_2.fit_transform(feature_used)

data_x_1 = pd.DataFrame(x_1, columns=feature_used.columns)
data_x_2 = pd.DataFrame(x_2, columns=feature_used.columns)


In [8]:
def scores(obj, Predict, Feature2, Label2):
    print('Accuracy   on test set: {:.3f}'.format(obj.score(Feature2, Label2)))
    print('F1_score   on test set: {:.3f}'.format(f1_score(Label2, Predict, average='macro')))
    print('Precision  on test set: {:.3f}'.format(precision_score(Label2, Predict, average='macro')))
    print('Recall     on test set: {:.3f}'.format(recall_score(Label2, Predict, average='macro')))

# Split Data 80-20

# Logistic Regression

In [9]:
# Training menggunakan data hasil scalling dengan Standard Scaler
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)

# Logistic Regression
print("Menggunakan Standard Scaler")
logReg = LogisticRegression(tol=0.001)
logReg.fit(feature_train,label_train)
cross_val= cross_val_score(logReg, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= logReg.predict(feature_test)
scores(logReg, prediction, feature_test, label_test)

print("-----------------------------------------------------------------------------------------------------------------------------")
print()

print("Menggunakan PowerTransform Scaller")
# Training menggunakan data hasil scalling dengan Power Transform Scaler
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20,random_state=4)
logReg = LogisticRegression(tol=0.001)
logReg.fit(feature_train,label_train)
cross_val_= cross_val_score(logReg, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= logReg.predict(feature_test)
scores(logReg, prediction, feature_test, label_test)


Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.869
F1_score   on test set: 0.866
Precision  on test set: 0.863
Recall     on test set: 0.871
-----------------------------------------------------------------------------------------------------------------------------

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.918
F1_score   on test set: 0.915
Precision  on test set: 0.918
Recall     on test set: 0.912


# MLP Classifier

In [10]:
print("Menggunakan Standard Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
mlc = MLPClassifier(hidden_layer_sizes=50, activation='identity',solver='lbfgs',batch_size='auto', learning_rate_init=0.0001, max_iter=10000,early_stopping=False)
mlc.fit(feature_train,label_train)
cross_val_= cross_val_score(mlc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= mlc.predict(feature_test)
scores(mlc, prediction, feature_test, label_test)

print("-----------------------------------------------------------------------------------------------------------------------------")
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
mlc= MLPClassifier(hidden_layer_sizes=50, activation='identity',solver='lbfgs',batch_size='auto', learning_rate_init=0.0001, max_iter=10000, early_stopping=False)
mlc.fit(feature_train,label_train)
cross_val_= cross_val_score(mlc, feature_train, label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= mlc.predict(feature_test)
scores(mlc, prediction, feature_test, label_test)

Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.869
F1_score   on test set: 0.866
Precision  on test set: 0.863
Recall     on test set: 0.871
-----------------------------------------------------------------------------------------------------------------------------

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.918
F1_score   on test set: 0.915
Precision  on test set: 0.918
Recall     on test set: 0.912


# Random Forest Classifier

In [11]:
estimator =[5,10,20,30,50,75,100,150]
for i in estimator:
    print("Menggunakan Standard Scaler")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Untuk Estimator "+str(i))
    rfc = RandomForestClassifier(i,criterion='gini',bootstrap=True,class_weight='balanced_subsample')
    rfc.fit(feature_train,label_train)
    cross_val_= cross_val_score(rfc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= rfc.predict(feature_test)
    scores(rfc, prediction, feature_test, label_test)
    print()

    print("Menggunakan PowerTransform Scaller")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    rfc = RandomForestClassifier(i,criterion='gini',bootstrap=True,class_weight='balanced_subsample')
    rfc.fit(feature_train,label_train)
    cross_val_= cross_val_score(rfc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= rfc.predict(feature_test)
    scores(rfc, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan Standard Scaler
Untuk Estimator 5
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.802
Precision  on test set: 0.805
Recall     on test set: 0.815

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.787
F1_score   on test set: 0.778
Precision  on test set: 0.780
Recall     on test set: 0.777

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan Standard Scaler
Untuk Estimator 10
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.918
F1_score   on test set: 0.916
Precision  on test set: 0.914
Recall     on test set: 0.918

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test s

# Gausssian-Process Classifier

In [12]:
print("Menggunakan Standard Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
kernel = 1.0 * RBF(1.0)
gpc = GaussianProcessClassifier(kernel=kernel, max_iter_predict=100, optimizer='fmin_l_bfgs_b')
gpc.fit(feature_train,label_train)
cross_val_= cross_val_score(gpc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gpc.predict(feature_test)
scores(gpc, prediction, feature_test, label_test)
print()

print("Menggunakan Power Transform Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
kernel = 1.0 * RBF(1.0)
gpc = GaussianProcessClassifier(kernel=kernel, max_iter_predict=100,optimizer='fmin_l_bfgs_b')
gpc.fit(feature_train,label_train)
cross_val_= cross_val_score(gpc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gpc.predict(feature_test)
scores(gpc, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.885
F1_score   on test set: 0.881
Precision  on test set: 0.883
Recall     on test set: 0.878

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.902
F1_score   on test set: 0.898
Precision  on test set: 0.898
Recall     on test set: 0.898



# KNN Classifier

In [13]:
numbers = 16
for i in range(numbers):
    if i %2 != 0 :
        print("Menggunakan nilai k : "+str(i))
        print("Menggunakan Standard Scaler")
        feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
        knn = KNeighborsClassifier(i,weights='distance',algorithm='auto')
        knn.fit(feature_train,label_train)
        cross_val_= cross_val_score(knn, feature_train,label_train, cv=5)
        print("Cross Validation Score : "+str(cross_val))
        prediction= knn.predict(feature_test)
        scores(knn, prediction, feature_test, label_test)
        print()

        print("Menggunakan Power Transform Scaler")
        feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
        knn = KNeighborsClassifier(i,weights='distance',algorithm='auto')
        knn.fit(feature_train,label_train)
        cross_val_= cross_val_score(knn, feature_train,label_train, cv=5)
        print("Cross Validation Score : "+str(cross_val))
        prediction= knn.predict(feature_test)
        scores(knn, prediction, feature_test, label_test)
        print()
        print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan nilai k : 1
Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.770
F1_score   on test set: 0.767
Precision  on test set: 0.767
Recall     on test set: 0.775

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.820
F1_score   on test set: 0.819
Precision  on test set: 0.834
Recall     on test set: 0.841

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan nilai k : 3
Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.799
Precision  on test set: 0.797
Recall     on test set: 0.803

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy 

# Adaboost Classifier

In [14]:
estimator =[5,10,20,30,50,75,100,150]
for i in estimator:
    print("Menggunakan Standard Scaler")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Untuk Estimator "+str(i))
    adc = AdaBoostClassifier(n_estimators=i,learning_rate=0.0001, random_state=1,algorithm='SAMME')
    adc.fit(feature_train,label_train)
    cross_val_= cross_val_score(adc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= adc.predict(feature_test)
    scores(adc, prediction, feature_test, label_test)
    print()

    print("Menggunakan PowerTransform Scaller")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    adc = AdaBoostClassifier(n_estimators=i,learning_rate=0.0001, random_state=1,algorithm='SAMME')
    adc.fit(feature_train,label_train)
    cross_val_= cross_val_score(adc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= adc.predict(feature_test)
    scores(adc, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan Standard Scaler
Untuk Estimator 5
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.797
Precision  on test set: 0.797
Recall     on test set: 0.797

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.797
Precision  on test set: 0.797
Recall     on test set: 0.797

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan Standard Scaler
Untuk Estimator 10
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.797
Precision  on test set: 0.797
Recall     on test set: 0.797

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test s

# Decision Tree Classifier

In [15]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
dtc=DecisionTreeClassifier(criterion='entropy',splitter='best',class_weight='balanced')
dtc.fit(feature_train,label_train)
cross_val_= cross_val_score(dtc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= dtc.predict(feature_test)
scores(dtc, prediction, feature_test, label_test)
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
dtc = DecisionTreeClassifier(criterion='entropy',splitter='best',class_weight='balanced')
dtc.fit(feature_train,label_train)
cross_val_= cross_val_score(dtc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= dtc.predict(feature_test)
scores(dtc, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.799
Precision  on test set: 0.797
Recall     on test set: 0.803

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.803
F1_score   on test set: 0.797
Precision  on test set: 0.797
Recall     on test set: 0.797



# Gaussian Naive-Bayes

In [16]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
gnb=GaussianNB()
gnb.fit(feature_train,label_train)
cross_val_= cross_val_score(gnb, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gnb.predict(feature_test)
scores(gnb, prediction, feature_test, label_test)
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
gnb=GaussianNB()
gnb.fit(feature_train,label_train)
cross_val_= cross_val_score(gnb, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gnb.predict(feature_test)
scores(gnb, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.836
F1_score   on test set: 0.828
Precision  on test set: 0.834
Recall     on test set: 0.824

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.885
F1_score   on test set: 0.882
Precision  on test set: 0.880
Recall     on test set: 0.884



# SVC

In [17]:
kernels=['poly','linear','rbf']
for i, value in enumerate(kernels):
    print("Menggunakan kernel : "+str(value))
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Menggunakan Standard Scaler")
    svl=SVC(kernel= value, C=1000,gamma='scale',tol=0.0001,class_weight='balanced',degree=3,coef0=1)
    svl.fit(feature_train,label_train)
    cross_val_= cross_val_score(svl, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= svl.predict(feature_test)
    scores(svl, prediction, feature_test, label_test)
    print()

    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    print("Menggunakan Power Transform Scaler")
    svl=SVC(kernel= value, C=1000,gamma='scale',tol=0.0001,class_weight='balanced',degree=3,coef0=1)
    svl.fit(feature_train,label_train)
    cross_val_= cross_val_score(svl, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= svl.predict(feature_test)
    scores(svl, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan kernel : poly
Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.770
F1_score   on test set: 0.767
Precision  on test set: 0.767
Recall     on test set: 0.775

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.770
F1_score   on test set: 0.770
Precision  on test set: 0.780
Recall     on test set: 0.787

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan kernel : linear
Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.902
F1_score   on test set: 0.899
Precision  on test set: 0.897
Recall     on test set: 0.904

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Acc

# LinearSVC

In [18]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
linearsvc = LinearSVC(class_weight='balanced',max_iter=1000)
linearsvc.fit(feature_train,label_train)
cross_val_= cross_val_score(linearsvc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= linearsvc.predict(feature_test)
scores(svl, prediction, feature_test, label_test)
print()

feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
print("Menggunakan Power Transform Scaler")
linearsvc = LinearSVC(class_weight='balanced',max_iter=1000)
linearsvc.fit(feature_train,label_train)
cross_val_= cross_val_score(linearsvc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= linearsvc.predict(feature_test)
scores(svl, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.738
F1_score   on test set: 0.866
Precision  on test set: 0.863
Recall     on test set: 0.871

Menggunakan Power Transform Scaler
Cross Validation Score : [0.71428571 0.81632653 0.85416667 0.79166667 0.875     ]
Accuracy   on test set: 0.770
F1_score   on test set: 0.898
Precision  on test set: 0.898
Recall     on test set: 0.898



# Kesimpulan

Pada percobaan diatas, menggunakan dataset 'Diabetes'. digunakan 10 algoritma untuk mengklasifikasi data menjadi 2 kelas
yaitu 0 = 'Tidak terkena Diabetes' , 1 = 'Terkena Diabetes' .

Dataset tersebut di-scaling dengan 2 scaler, yaitu 'Standard scaler' dan 'Power Transform'.

* Algoritma LogisticRegression mendapat nilai akurasi dan precision

    Standard scaler : 0.869 dan 0.863 , Power Transform : 0.918 dan 0,918

* Algoritma MLP Classifier mendapat nilai akurasi dan precision

    Standard scaler : 0.869 dan 0.863 , Power Transform : 0.918 dan 0.918

* Algoritma RandomForest Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai estimator 30, dengan :

    Standard scaler : 0.932 dan 0.930 , Power Transform : 0.852 dan 0.847

* Algoritma KNN Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai k = 13, dengan :

    Standard scaler : 0.820 dan 0.15 , Power Transform : 0.903 dan 0.902

* Algoritma Adaboost Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai estimator 5, dengan :

    Standard scaler : 0.803 dan 0.797 , Power Transform : 0.803 dan 0.797

* Algoritma Decision Tree Classifier mendapat nilai akurasi dan precision

    Standard scaler : 0.803 dan 0.797 , Power Transform : 0.803 dan 0.797

* Algoritma Gaussian Naive-Bayes mendapat nilai akurasi dan precision

    Standard scaler : 0.830 dan 0.834 , Power Transform : 0.885 dan 0.880

* Algoritma SVC yang mendapat nilai terbaik menggunakan kernel 'linear'
dengan nilai akurasi dan precision

    Standard scaler : 0.902 dan 0.897 , Power Transform : 0.885 dan 0.880

* Algoritma LinearSVC mendapat nilai akurasi dan precision

    Standard scaler : 0.738 dan 0.863 , Power Transform : 0.770, 0.898

Dari hasil percobaan diatas, algoritma RandomForest Classifier, menggunakan data yang
di-scaling dengan Standard scaler mendapat hasil akurasi yang terbaik sebesar 0.932

Maka, dapat disimpulkan bahwa penggunaan Standard scaler pada dataset ini sudah mencukupi
untuk bisa mendapat nilai akurasi dan precision yang cukup baik pada kebanyakan algoritma yang digunakan. 
Tapi, dalam beberapa algoritma akan lebih baik hasilnya jika menggunakan Power Transform scaler.

Hasil klasifikasi yang didapatkan oleh model algoritma dapat dipengaruhi oleh banyak faktor,
salah satunya adalah penggunaan Hyperparameter pada tiap algoritma dan keseimbangan jumlah sebaran data pada label.
jika seimbang, maka pengklasifikasian data akan terjadi merata / tidak condong pada 1 kelas.